### ISSUE TITLE
Grammar for constObjectExpression is ambiguous.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#22393]
The grammar for constObjectExpression is:

```
constObjectExpression:
  const type (’.’ identifier)? arguments
;

type:
  typeName typeArguments?
;

typeName:
  qualified
;

qualified:
  identifier ('.' identifier)?
;
```

This means that &quot;const foo.foo()&quot; can be parsed in two ways:
- one where &quot;type&quot; -&gt; typeName -&gt; qualified -&gt; identifier -&gt; foo and &quot;identifier&quot; -&gt; foo in constObjectExpression, and
- one where &quot;type&quot; -&gt; typeName -&gt; qualified -&gt; identifier '.' identifier -&gt; foo '.' foo and ('.' identifier)? is empty in constObjectExpression.

The specification does not say how to disambiguate, only suggestively calls one part &quot;T&quot;, but parsing &quot;prefix.ClassName&quot; by the former strategy would cause the semantics to require a compile time error, since &quot;prefix&quot; is not the name of a type.

**Comment by eernstg** [comment]
No milestone now: This correction is part of a larger update of grammar related parts of the specification, based on [Dart.g](https://github.com/dart-lang/sdk/blob/master/docs/language/Dart.g). Note that the fact that Dart.g is not ambiguous may imply that the issue has already been resolved there.

**Comment by eernstg** [comment]
New stuff on this topic: About https://dart-review.googlesource.com/c/sdk/+/51323/46/docs/language/dartLangSpec.tex#5204:

> I'm slightly suspicious about this, because how do you disambiguate `id1.id2` before finding out what `id1` is? 
> Could I argue that `C.y` should be treated by the $T$ case and it's a compile-time error because it's not a class?
> Or that `p.C` should be parsed as `$T$.id` and it's a compile-time error because p (a prefix) is not a class?
>
> (In other words: We have ambiguous syntax, and we need to disambiguate *before* we start saying that something is a compile-time error, otherwise a valid implementation can validly make any program non-deterministically valid/invald).
>
> This is where "if $c$ denotes a constructor" would be a good abstraction.

**Comment by eernstg** [comment]
In Dart.g the rule for `constObjectExpression` has right hand side `CONST constructorDesignation arguments` where `constructorDesignation` is `qualified | typeName typeArguments ('.' identifier)?`, which resolves the ambiguity. So this is now a matter of transferring fixes from Dart.g to the spec.

**Comment by eernstg** [comment]
Fixed in dart-lang/language#629.
